Update update-api-with-oas.md (#2382)
TykTechnologies/tyk-docs: tyk-docs/content/getting-started/using-oas-definitions/update-api-with-oas.md

---
title: "Update existing Tyk API with changes in your OpenAPI definition"
date: 2022-07-13
tags: ["Tyk Tutorials OpenAPI", "Getting Started OpenAPI", "First API OpenAPI", "Tyk Cloud OpenAPI", "Tyk Self-Managed OpenAPI", "Tyk Open Source OpenAPI", "Updating API with OAS", "Update OpenAPI definition", "Import an OpenAPI file to update an existing API definition"]

> description: "Updating Tyk OAS API definition following in your OpenAPI defintion"

Reviewer: ```suggestion
description: "Updating Tyk OAS API following a change in your OpenAPI definition"
```

### Introduction

> While developing APIs, as developers, we constantly need to update the OpenAPI Specification that is either generated from our codebase, or built using some API design tools. Tyk allows you to update the Tyk API definition with fresh copy of your OpenAPI Specification using PATCH HTTP method.

Author: ```suggestion
While developing APIs, as developers, we constantly need to update the OpenAPI Specification that is either generated from our codebase or built using some API design tools. Tyk allows you to update the Tyk OAS API definition with a fresh copy of your OpenAPI Specification using the PATCH HTTP method.
```

> While developing APIs, as developers, we constantly need to update the OpenAPI Specification that is either generated from our codebase, or built using some API design tools. Tyk allows you to update the Tyk API definition with fresh copy of your OpenAPI Specification using PATCH HTTP method.

Reviewer: Looks good. 👍

> While developing APIs, as developers, we constantly need to update the OpenAPI Specification that is either generated from our codebase or built using some API design tools. Tyk allows you to update the Tyk OAS API definition with a fresh copy of your OpenAPI Specification using the PATCH HTTP method.

Author: @andyo-tyk your message is gone, please re-review this bit. 
I used "some" since wasn't sure if I should name a few tools like swagger UI and stoplight or not. What do you think is best? essentially we want to say it's outside tyk's domain

### Open Source

> ### Tutorial: Update Tyk OAS API definition with an updated OpenAPI Specification

Author: ```suggestion
### Tutorial: Update Tyk OAS API definition with an updated OpenAPI Specification
### Tutorial: Update a Tyk OAS API with an updated OpenAPI Specification
```

> ### Tutorial: Update Tyk OAS API definition with an updated OpenAPI Specification

Author: I think we do since it's not an API, it's an API definition but I removed it

> ### Tutorial: Update Tyk OAS API definition with an updated OpenAPI Specification

Reviewer: On reflection, you're right, we do need something at the end of the sentence - I understand it's an OAS document (https://spec.openapis.org/oas/v3.1.0#openapi-document)... and I think we don't need it after Tyk OAS API, because we're actually talking about updating the API on the Gateway. 🤦 

```suggestion
### Tutorial: Update a Tyk OAS API with an updated OpenAPI Specification document 
```

#### Make sure you know your Tyk API secret

Your Tyk Gateway API secret is stored in your `tyk.conf` file, the property is called `secret`. You will need to use this as a header called `x-tyk-authorization` to make calls to the Gateway API.

#### Create an API

> For the sake of this tutorial, create a new API by sending this Tyk OAS API Definition [https://bit.ly/39tnXgO](https://bit.ly/39tnXgO) to the Gateway API endpoint (this is an example that contains the very minimal required fields):

Reviewer: ```suggestion
For the sake of this tutorial, create a new API by sending this [Tyk OAS API Document](https://bit.ly/39tnXgO) to the Gateway API endpoint (this is an example that contains the very minimal required fields):
```

> For the sake of this tutorial, create a new API by sending this Tyk OAS API Definition [https://bit.ly/39tnXgO](https://bit.ly/39tnXgO) to the Gateway API endpoint (this is an example and contains the very minimun required fields):

Author: ```suggestion
For the sake of this tutorial, create a new API by sending this Tyk OAS API Definition [https://bit.ly/39tnXgO](https://bit.ly/39tnXgO) to the Gateway API endpoint (this is an example that contains the very minimal required fields):
```

curl -H "x-tyk-authorization: {your-secret}" -s http://{your-tyk-host}:{port}/tyk/reload/group
```

#### Update your OpenAPI Specification outside Tyk

> Now let's assume you made some change in your OpenAPI Specification using some tool outside Tyk's domain. It could be for example adding a new path, changing description or any other change.

Reviewer: Perhaps there's a missing "the" between "changing" and "description"?

> Now let's assume you made some change in your OpenAPI Specification using some tool outside Tyk's domain. It could be for example adding a new path, changing description or any other change.

Author: ```suggestion
Now let's assume you made some changes in your OpenAPI Specification using some tool outside Tyk's domain. It could be for example adding a new path, changing the description or any other change.
```

> Now let's assume you made some change in your OpenAPI definition (as mentioned above, from code or a tool, outside Tyk's domain). The change could be for  adding a new path, changing a description or anything that changes the definition.

Reviewer: ```suggestion
Now let's assume you made a change in your OpenAPI definition (as mentioned above, from code or a tool, outside Tyk's domain). The change could be adding a new path, changing a description or anything that changes the definition of the API.
```

> In this case we added a new endpoint, `POST /pet`, with a schema that validates the payload it receives (`requestBody.content.application/json.schema`) and a new security scheme (The OpenAPI Specification is in the code snippet in the next section, just to avoid repeatition) 

Author: spot on!

> In this case we added a new endpoint, `POST /pet`, with a schema that validates the payload it receives (`requestBody.content.application/json.schema`) and a new security scheme (The OpenAPI Specification is in the code snippet in the next section, just to avoid repeatition) 

Author: ```suggestion
In this example we added a new endpoint, `POST /pet`, with a schema that validates the payload it receives (`requestBody.content.application/json.schema`) and a new security scheme (The OpenAPI Specification is in the code snippet in the next section, just to avoid repetition) 
```

> In this example we added a new endpoint, `POST /pet`, with a schema that validates the payload it receives (`requestBody.content.application/json.schema`) and a new security scheme (The OpenAPI definition is in the code snippet in the next section, just to avoid repetition) 

Reviewer: ```suggestion
In this example we added a new endpoint, `POST /pet`, with a schema that validates the payload it receives (`requestBody.content.application/json.schema`) and a new security scheme (the OpenAPI definition is in the code snippet in the next section, just to avoid repetition) 
```

> #### Update the Tyk API definition using just your updated OpenAPI Specification

Author: ```suggestion
#### Update the Tyk 
OAS API using just your updated OpenAPI Specification
```

> Tyk allows you to update a Tyk OAS API definition by providing only your OpenAPI Specification using `PATH` request.

Reviewer: 1) Again, I would remove the word "definition"
2) There's a typo: PATH -> PATCH
3) Again, I would add "the" before PATCH

Author: 1. In the title we can be shorter but in the explanation, we should be accurate, why remove the word "definition"?

> Tyk allows you to update a Tyk OAS API definition by providing only your OpenAPI Specification using `PATH` request.

Author: ```suggestion
Tyk allows you to update a Tyk OAS API definition by providing only your OpenAPI Specification using the `PATCH` request.
```

> Tyk allows you to update a Tyk OAS API definition by providing only your OpenAPI definition using `PATH` request.

Reviewer: ```suggestion
You can update your Tyk OAS API definition very simply - by providing your OpenAPI definition as an OpenAPI document with the `PATH` request.
```

> You can update your Tyk OAS API definition very simply - by providing your OpenAPI definition as an OpenAPI document with the `PATH` request.

Author: ```suggestion
You can update your Tyk OAS API definition very simply - by providing your OpenAPI definition as an OpenAPI document with the `PATCH` request.
```

> Tyk will update the OpenAPI Specification section in the Tyk OAS API dfintion.

Reviewer: Typo: dfintion
(though I would just delete the word :)

> Tyk will update the OpenAPI Specification section in the Tyk OAS API dfintion.

Author: ```suggestion
As a result, Tyk will replace the existing OpenAPI Specification in the Tyk OAS API with the new OpenAPI Specification (from the payload).
```

> Tyk will update the OpenAPI definition section in the Tyk OAS API dfintion.

Reviewer: ```suggestion
Tyk will use the content of the OpenAPI document to update just the OpenAPI section in the Tyk OAS API definition.
```

Once you have created your API, you will need to either restart the Tyk Gateway, or issue a hot reload command with the following curl command:

```curl
curl -H "x-tyk-authorization: {your-secret}" -s http://{your-tyk-host}:{port}/tyk/reload/group
```

#### Protect your API based on the OpenAPI definition

> Previously you updated the Tyk OAS API Definition with a new OpenAPI definition, which is describing a security mechanism. In order for Tyk Gateway to use this security mechanism and start protecting the API, the authentication mechanism needs to be *enabled* within Tyk section in the Tyk OAS API definiton.

Reviewer: ```suggestion
Previously you updated the Tyk OAS API definition with a new OpenAPI definition, that describes a new security mechanism. In order for Tyk Gateway to start protecting the API using this authentication mechanism, it needs to be *enabled* within the Tyk section of the Tyk OAS API definition.
```

> Previously you updated the Tyk OAS API Definition with a new OpenAPI Specification, which is describing a security mechanism. In order for Tyk Gateway to use this security mechanism and start protecting the API, the authentication mechanism needs to be *enabled* within Tyk section in the Tyk OAS API definiton.

Reviewer: 1) I'd remove "defintion"
2) "In order to" or "In order for" is unnecessarily wordy: perhaps just say "For Tyk Gateway to use this security mechanism..."
3) missing "the" between "within" and "Tyk"
4) Typo: definiton -> definition

> Previously you updated the Tyk OAS API Definition with a new OpenAPI Specification, which is describing a security mechanism. In order for Tyk Gateway to use this security mechanism and start protecting the API, the authentication mechanism needs to be *enabled* within Tyk section in the Tyk OAS API definiton.

Author: ```suggestion
Previously you updated the Tyk OAS API Definition with a new OpenAPI Specification, which is describing a security mechanism. For Tyk Gateway to use this security mechanism and start protecting the API, the authentication mechanism needs to be *enabled* within the Tyk section in the Tyk OAS API definition.
```

> For that, together with the `PATCH` request you just performed, add the authentication query parameter `authentication=true`, that tells Tyk to automatically enable authentication, based on the settings in the OAS definition.

Reviewer: ```suggestion
To do this you would add the query parameter `authentication=true` to the `PATCH` request you just performed; this tells Tyk to automatically enable authentication, based on the settings in the OpenAPI definition.
```

```
#### Explicitly allow access to documented endpoints

> While updating a Tyk API OAS Definition using just the OpenAPI definition, you can also give instructions to Tyk Gateway to explicitly allow access just to paths that are documented in the OAS API Definition. For that, we have to pass the `allowList` query parameter together with our payload.

Reviewer: ```suggestion
Tyk Gateway's allow list function explicitly allows access just to paths that are documented in the Tyk OAS API definition. You can enable this when updating a Tyk API OAS definition using the `PATCH` method by passing the `allowList` query parameter with the payload.
```

> While updating a Tyk API OAS Definition using just the OpenAPI Specification, you can also give instructions to Tyk Gateway to explicitly allow access just to paths that are documented in the OAS API Definition. For that, we have to pass the `allowList` query parameter together with our payload.

Reviewer: Perhaps reword this to:
You can instruct the Tyk Gateway to allow access only to paths (endpoints) explicitly documented in the Tyk OAS API using the allow list functionality. If you pass the `allowList` query parameter during the `PATCH` operation, the allow list will be automatically configured according to the OpenAPI Specification in the payload.

 --> this perhaps gives the opportunity to add a link (to here? https://tyk.io/docs/tyk-apis/tyk-gateway-api/api-definition-objects/ip-whitelisting/)

Author: ```suggestion
You can instruct the Tyk Gateway to allow access only to paths (endpoints) explicitly documented in the OpenAPI Specification, using the allow list functionality. If you pass the `allowList` query parameter during the PATCH operation, the allow list will be automatically configured.
```

Author: 1. Changed it a bit as we were repeating ourselves.
2. This link shows fields of classic API def. Do we have OAS API def?

    }
}
```

#### What did you just do?

> You have demonstrated that by using theOpenAPI definition, which can be either generated from your source code or created as part of design first approach, you can update and configure the Tyk OAS API Definition accordingly.

Reviewer: ```suggestion
You have demonstrated that by using just the OpenAPI definition, which can be either generated from your source code or created as part of design first approach using other tools, you can update and configure an existing Tyk OAS API Definition.
```

> You have demonstrated that by using theOpenAPI specification, which can be either generated from your source code or created as part of design first approach, you can update and configure the Tyk OAS API Definition accordingly.

Reviewer: Missing a space in "the OpenAPI specification"

Perhaps "a" or "your" before "design first approach"?

As above, I'd remove the word "Definition".
